### PR TITLE
DEV-2456: support client side timeout

### PR DESCRIPTION
**Pull request by h12w**
1. Set default client timeout to 10s and allow users to modify it through command line options and environmental variables
2. `storageos logs -f` will not be affected by the value because it uses its own client
3. `storageos node rm` will have minimum timeout of 60s.

CI:
http://ci.storageos.net/job/product-build/job/feature%252FDEV-2456-default-timeout/

Manually tried most sub commands on my demo-env. All can be completed within 1s except volume create and deletion. Volume creation takes less than 2s. volume deletion takes less than 10s (default 60s).

### CODE REVIEW

**Comment by h12w** [comment]
http://ci.storageos.net/job/product-build/job/feature%252FDEV-2456-default-timeout-client/

File: cli/command/cli.go [hunk 1]
```diff
@@ -50,6 +52,7 @@ type StorageOSCli struct {
 	client          *api.Client
 	hasExperimental bool
 	defaultVersion  string
+	timeout         int
```
Comment (domodwyer) [inline cli/command/cli.go]: Why store this as an int?

It's idiomatic to store time durations as a `time.Duration` in Go, as there is no question regarding the unit of duration (seconds, milliseconds, minutes). All `time.Duration` values are stored in nanoseconds, and it has a bunch of helpful methods to convert to other bases.

```suggestion
	timeout         time.Duration
```

File: cli/command/cli.go [hunk 2]
```diff
@@ -232,6 +250,15 @@ func getDiscovery(discoveryFlag string) string {
 	return discoveryHost
 }
 
+func getTimeout() (int, error) {
+	timeoutStr := os.Getenv(cliconfig.EnvStorageOSTimeout)
+
+	if timeoutStr == "" {
+		return cliconfig.DefaultTimeout, nil
+	}
+	return strconv.Atoi(timeoutStr)
```
Comment (domodwyer) [inline cli/command/cli.go]: Combined with using a `time.Duration`, we can give the user a nicer experience by accepting different units for free:
```suggestion
	return time.ParseDuration(timeoutStr)
```

This allows the user to specify values like "5s" or "1m" - however if they specify "5" it'll default to 5 nanoseconds, so we should have some logic to convert values without a unit to seconds.

File: cli/flags/common.go [hunk 3]
```diff
@@ -61,6 +62,8 @@ func (commonOpts *CommonOptions) InstallFlags(flags *pflag.FlagSet) {
 
 	flags.StringVarP(&commonOpts.Username, "username", "u", "", fmt.Sprintf(`API username (will override %s env variable value)`, cliconfig.EnvStorageosUsername))
 	flags.StringVarP(&commonOpts.Password, "password", "p", "", fmt.Sprintf(`API password (will override %s env variable value)`, cliconfig.EnvStorageosPassword))
+
+	flags.IntVarP(&commonOpts.Timeout, "timeout", "t", 0, fmt.Sprintf(`client timeout in seconds (will override %s env variable value if set), default 10s`, cliconfig.EnvStorageOSTimeout))
```
Comment (domodwyer) [inline cli/flags/common.go]: I'm pretty sure the 4th field is the default value, and I would imagine the default value is included in the output automatically

```suggestion
flags.IntVarP(&commonOpts.Timeout, "timeout", "t", cliconfig.DefaultTimeout, fmt.Sprintf(`client timeout in seconds (will override %s env variable value if set)`, cliconfig.EnvStorageOSTimeout))
```

However we should swap this to:

```suggestion
	help := fmt.Sprintf("client timeout in seconds (will override %s env variable value if set)", cliconfig.EnvStorageOSTimeout)
	flags.DurationVarP(p *time.Duration, "timeout", "t", cliconfig.DefaultTimeout, help)
```

File: cli/command/cli.go (hunk 1 above)
Comment (h12w) [reply cli/command/cli.go]: done

File: cli/command/cli.go (hunk 2 above)
Comment (h12w) [reply cli/command/cli.go]: done

File: cli/flags/common.go (hunk 3 above)
Comment (h12w) [reply cli/flags/common.go]: The default value is from environmental variable, so it should not be set from command line argument. Same logic as controlplane.

Comment (domodwyer) [reply cli/flags/common.go]: Can't we make use of `flags.Changed()` like in the func below? Seems like we're working against the CLI library here, that's all.

Comment (h12w) [reply cli/flags/common.go]: done

File: cli/flags/common.go [hunk 4]
```diff
@@ -24,6 +25,11 @@ const (
 	FlagTLSVerify = "tlsverify"
 )
 
+const (
+	// FlagTimeout is timeout flag
+	FlagTimeout = "timeout"
```
Comment (domodwyer) [inline cli/flags/common.go]: Err can we put this with the others in the `const` block above?

File: cli/flags/common.go [hunk 5]
```diff
@@ -61,6 +68,8 @@ func (commonOpts *CommonOptions) InstallFlags(flags *pflag.FlagSet) {
 
 	flags.StringVarP(&commonOpts.Username, "username", "u", "", fmt.Sprintf(`API username (will override %s env variable value)`, cliconfig.EnvStorageosUsername))
 	flags.StringVarP(&commonOpts.Password, "password", "p", "", fmt.Sprintf(`API password (will override %s env variable value)`, cliconfig.EnvStorageosPassword))
+
+	flags.DurationVarP(&commonOpts.Timeout, FlagTimeout, "t", cliconfig.DefaultTimeout, fmt.Sprintf(`client timeout in seconds (will override %s env variable value if set), default 10s`, cliconfig.EnvStorageOSTimeout))
```
Comment (domodwyer) [inline cli/flags/common.go]: We can drop the explicit `default 10s` in the description now 👍

Comment (h12w) [reply cli/flags/common.go]: done

File: cli/flags/common.go (hunk 4 above)
Comment (h12w) [reply cli/flags/common.go]: done

**Comment by h12w** [comment]
CI testing link is attached in the JIRA ticket.

File: cli/command/cli.go (hunk 2 above)
Comment (domodwyer) [reply cli/command/cli.go]: Given this needs a duration unit, I think we should either implement the above (defaulting to seconds if no unit is provided) or include (a subset of) the acceptable units in the help text.

After that, looks good to me.

Comment (domodwyer) [reply cli/command/cli.go]: Or maybe just intercepting the error and printing something more helpful? I'll let you see what works best 👍

Comment (h12w) [reply cli/command/cli.go]: Added some examples in the help text.

**Pull request merged**